components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java

```diff
@@ -183,4 +189,81 @@ public void writeJSON(Patient patient, JSONObject json, Collection<String> selec
             }
         }
     }
+
+    @Override
+    public PatientData<Map<String, String>> readJSON(JSONObject json)
+    {
+        Object testGeneJson = json.get(this.getJsonPropertyName());
+        if (testGeneJson == null) {
+            return null;
+        }
+
+        try {
+            JSONArray genesJson = json.getJSONArray(this.getJsonPropertyName());
+            List<Map<String, String>> allGenes = new LinkedList<Map<String, String>>();
+            for (Object geneJsonUncast : genesJson) {
+                JSONObject geneJson = JSONObject.fromObject(geneJsonUncast);
+                Map<String, String> singleGene = new LinkedHashMap<String, String>();
+                for (String property : this.getProperties()) {
+                    if (geneJson.has(property)) {
+                        String field = geneJson.getString(property);
+                        if (field != null) {
+                            singleGene.put(property, field);
+                        }
+                    }
+                }
+                if (!singleGene.isEmpty()) {
+                    allGenes.add(singleGene);
+                }
+            }
+
+            if (allGenes.isEmpty()) {
+                return null;
+            } else {
+                return new IndexedPatientData<Map<String, String>>(getName(), allGenes);
+            }
+        } catch (Exception e) {
+            this.logger.error("Could not load genes from JSON", e.getMessage());
+        }
+        return null;
+    }
+
+    @Override
+    public void save(Patient patient)
+    {
+        try {
+            XWikiDocument doc = (XWikiDocument) this.documentAccessBridge.getDocument(patient.getDocument());
+            if (doc == null) {
+                throw new NullPointerException(ERROR_MESSAGE_NO_PATIENT_CLASS);
+            }
+
+            PatientData<Map<String, String>> genes = patient.getData(this.getName());
+            if (!genes.isIndexed()) {
+                return;
+            }
+
+            XWikiContext context = (XWikiContext) this.execution.getContext().getProperty("xwikicontext");
+            Iterator<Map<String, String>> iterator = genes.iterator();
+            while (iterator.hasNext()) {
+                try {
+                    Map<String, String> gene = iterator.next();
+                    BaseObject xwikiObject = doc.newXObject(GENE_CLASS_REFERENCE, context);
```

Comment:
You're adding new objects, but what happens if the patient already had some objects? You'll get duplicates, or have deprecated objects.